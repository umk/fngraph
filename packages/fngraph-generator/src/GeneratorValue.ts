import { Context } from '@fngraph/data'

type GeneratorValue = {
  /**
   * An index of node in an ordered array of nodes that has
   * advanced the last time.
   */
  n: number
  /**
   * The contexts that has been accepted the last time by
   * respective nodes in an ordered collection of nodes. A node
   * can be undefined if the context declarations conflict with
   * declarations of its parent contexts.
   */
  contexts: GeneratorValueContexts
}

export type GeneratorValueContexts = Array<Context | undefined>

export default GeneratorValue
