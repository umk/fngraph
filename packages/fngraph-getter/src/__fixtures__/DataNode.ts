import { DataNode } from '@fngraph/generator'

export function createDataNode(overrides?: Partial<DataNode>): DataNode {
  return {
    incoming: [],
    outgoing: [],
    getter: jest.fn(),
    priority: 0,
    invert: false,
    getProperties: () => [],
    ...overrides,
  }
}
