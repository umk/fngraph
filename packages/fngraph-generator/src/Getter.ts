import { DataRecord } from '@fngraph/data'

import GeneratorContext from './GeneratorContext'
import GeneratorValue from './GeneratorValue'
import GetContextRecord from './GetContextRecord'
import IncomingMapper from './IncomingMapper'
import OutgoingMapper from './OutgoingMapper'
import PropertyRef from './PropertyRef'

/**
 * Creates a getter function with incoming and outgoing declarations
 * that respective mappers operate
 * @param incoming The incoming declarations mapper
 * @param outgoing The outgoing declarations mapper
 * @returns A getter function
 */
export type GetterFactory<P extends DataRecord, R extends DataRecord> = (
  incoming: IncomingMapper<P>,
  outgoing: OutgoingMapper<R>,
) => Getter

/**
 * A function, which iterates over the records and derives new
 * data against each of the record. Upon yield the getter must
 * update the indexes array in a context to reflect global unique
 * location of the resulting record.
 * @param properties A collection of properties to be collected from the getter output
 * @param records A collection of records over which to derive the values
 * @param context A node iterator context
 * @param getRecord A function which gets a context record from generator value
 * @param invert Indicates whether the getter output is inverted
 */
type Getter = (
  properties: Array<PropertyRef>,
  records: AsyncGenerator<GeneratorValue>,
  context: GeneratorContext,
  getRecord: GetContextRecord,
  invert: boolean,
) => AsyncGenerator<GeneratorValue>

/**
 * The default max number of items to be fed into a batched
 * getter.
 */
export const DEFAULT_MAX_BATCH = 10

/**
 * Default number of generator values preloaded into buffer
 * in order to extract unique records that fed into a batched
 * getter.
 */
export const DEFAULT_MAX_BUFFER = 1000

export default Getter
