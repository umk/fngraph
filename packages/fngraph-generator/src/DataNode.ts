import { DeclarationID } from '@fngraph/data'

import Getter from './Getter'
import PropertiesGetter from './PropertiesGetter'

type DataNode = {
  /**
   * Declarations which values the node depends on.
   */
  incoming: Array<DeclarationID>
  /**
   * Declarations which values are produced by the node.
   */
  outgoing: Array<DeclarationID>
  /**
   * A function that actually produces declarations based on
   * combinations of incoming declarations.
   */
  getter: Getter
  /**
   * Indicates whether node output is inverted. The inverted
   * nodes cannot have outgoing declarations as they just
   * serve as a predicate for incoming declarations.
   */
  invert: boolean
  /**
   * The priority the node is invoked. Higher number means the
   * node is invoked later.
   */
  priority: number
  /**
   * A function that produces a collection of properties to be
   * collected from the getter output based on other nodes inputs.
   */
  getProperties: PropertiesGetter
}

export default DataNode
