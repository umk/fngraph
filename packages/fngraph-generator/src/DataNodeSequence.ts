import { DeclarationID } from '@fngraph/data'

import DataNode from './DataNode'
import PriorityQueue from './PriorityQueue'

class DataNodeSequence<N extends DataNode = DataNode> {
  private constructor(
    public readonly nodes: Array<N>,
    /**
     * @internal
     */
    public readonly incomings: Map<DeclarationID, Array<N>>,
  ) {}
  static create<N extends DataNode>(nodes: Array<N>) {
    const incomings = getIncomingNodes(nodes)
    const n = getOrderedNodes(nodes, incomings)
    return new DataNodeSequence<N>(n, incomings)
  }
}

function getIncomingNodes<N extends DataNode>(nodes: Array<N>) {
  return nodes.reduce((prev, cur) => {
    for (const incoming of cur.incoming) {
      let current = prev.get(incoming)
      if (!current) prev.set(incoming, (current = []))
      current.push(cur)
    }
    return prev
  }, new Map<DeclarationID, Array<N>>())
}

function getOrderedNodes<N extends DataNode>(
  nodes: Array<N>,
  incomings: Map<DeclarationID, Array<N>>,
): Array<N> {
  const indexes = new Map(nodes.map((n) => [n, 0]))
  const pending = new PriorityQueue<N>((a, b) => a.priority - b.priority)
  nodes.forEach((n) => pending.enqueue(n))
  for (;;) {
    const current = pending.dequeue()
    if (!current) break
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const i = indexes.get(current)!
    if (i === nodes.length) {
      throw new Error('The nodes have recursive dependency')
    }
    for (const outgoing of current.outgoing) {
      const n = incomings.get(outgoing)
      if (n) {
        for (const node of n) {
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          if (i >= indexes.get(node)!) {
            indexes.set(node, i + 1)
            pending.enqueue(node)
          }
        }
      }
    }
  }
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return nodes.sort((a, b) => indexes.get(b)! - indexes.get(a)!)
}

export default DataNodeSequence
