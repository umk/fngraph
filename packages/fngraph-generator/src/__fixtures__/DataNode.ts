import DataNode from '../DataNode'

export function createDataNode(overrides?: Partial<DataNode>): DataNode {
  return {
    incoming: [],
    outgoing: [],
    getter: jest.fn(),
    priority: 0,
    invert: false,
    getProperties: () => [],
    ...overrides,
  }
}
