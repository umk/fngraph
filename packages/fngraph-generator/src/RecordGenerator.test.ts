import { DataRecord, Declaration, DeclarationID } from '@fngraph/data'

import DataNode from './DataNode'
import DataNodeSequence from './DataNodeSequence'
import GeneratorValue from './GeneratorValue'
import Getter from './Getter'
import { createRecordGenerator } from './RecordGenerator'
import { createDataNode } from './__fixtures__'

function createGetter(values: Array<DataRecord>): Getter {
  return async function* (
    _properties,
    records,
    context,
    getRecord,
    invert,
  ): AsyncGenerator<GeneratorValue> {
    if (invert) throw new Error('inversion is not supported')
    for await (const current of records) {
      context.advanceToValueGroup(current)
      const base = getRecord(current.contexts)
      if (base) {
        let n = current.n
        for (const v of values) {
          current.contexts[context.nodeIndex] = { ...base, ...v }
          yield { contexts: current.contexts, n }
          n = context.nodeIndex
        }
      }
    }
  }
}

describe('createRecordGenerator', () => {
  const declarationA = 'declaration A' as DeclarationID
  const declarationB = 'declaration B' as DeclarationID
  const declarationC = 'declaration C' as DeclarationID

  it("doesn't return anything if there are no nodes", async () => {
    const nodes: Array<DataNode> = []
    const declarations: Array<Declaration> = []

    const sequence = DataNodeSequence.create(nodes)

    const generator = createRecordGenerator(sequence, declarations)
    const records = generator()

    const { done } = await records.next()

    expect(done).toBe(true)
  })
  it('iterates through getter records', async () => {
    const nodeA = createDataNode({
      outgoing: [declarationA, declarationB],
      getter: createGetter([
        { [declarationA]: 1, [declarationB]: 2 },
        { [declarationA]: 10, [declarationB]: 20 },
      ]),
    })
    const nodeB = createDataNode({
      incoming: [declarationA, declarationB],
      outgoing: [declarationC],
      getter: createGetter([
        { [declarationC]: 3 },
        { [declarationC]: 30 },
        { [declarationC]: 300 },
      ]),
    })
    const sequence = DataNodeSequence.create([nodeA, nodeB])
    const generator = createRecordGenerator(sequence, [])
    const records: Array<DataRecord> = []
    for await (const record of generator()) records.push(record)
    expect(records).toMatchSnapshot()
  })
})
