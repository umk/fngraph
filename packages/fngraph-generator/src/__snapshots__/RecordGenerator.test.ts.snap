// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRecordGenerator iterates through getter records 1`] = `
[
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 3,
  },
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 30,
  },
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 300,
  },
  {
    "declaration A": 10,
    "declaration B": 20,
    "declaration C": 3,
  },
  {
    "declaration A": 10,
    "declaration B": 20,
    "declaration C": 30,
  },
  {
    "declaration A": 10,
    "declaration B": 20,
    "declaration C": 300,
  },
]
`;
